### ISSUE TITLE
Reset connection without reconnect

### ISSUE THREAD

**Issue by lucidfrontier45** [rethinkdb/rethinkdb#5491]
Hi all.

Since Python driver is currently not thread-safe as described [here] ( https://rethinkdb.com/api/python/connect/), I'm working on a rethinkdb connection pool [library](https://github.com/lucidfrontier45/RethinkPool).

Now I want to know if I can just **reset** a connection that may be use by `run` and not all the results were fetched without re-initiating another tcp handshake.

I checked `reconnect(noreply_wait=False)` but found there is explicit socket close.

Thank you.

**Comment by danielmewes** [comment]
Hi @lucidfrontier45 , resetting the connection (i.e. closing all open cursors) is currently not possible without a complete reconnect.

It's probably possible to add this functionality to the driver (or even the server), though I'm unsure how important this is.

**Comment by lucidfrontier45** [comment]
according to discussion #2427, especially this [comment](https://github.com/rethinkdb/rethinkdb/issues/2427#issuecomment-140289274), reusing connection can will give roughly twice performance for simple query. Though the connection overhead is not important for complex queries, I think it is not ignorable for relatively simple query used for building REST APIs.

**Comment by danielmewes** [comment]
@lucidfrontier45 I agree that this becomes relevant if you need to do it frequently.
Can I ask why you need to reset connections at all in your connection pool though?

**Comment by lucidfrontier45** [comment]
I am afraid that leaving or reusing a connection that has unfetched result data may cause problem.

My concern is

1. leaving unfetched data can waste memoey on both server and client.
2. reusing connection a connection with unfetched data can corrupt the following result data.

I know nothing about RethinkDB internal and this concern might be misdirected.
If so I can just reuse connections without resetting or reconnecting.

**Comment by danielmewes** [comment]
> leaving unfetched data can waste memoey on both server and client.

That is true, but only for as long as the cursor that requested the data is still around. If you close the cursor, or it gets garbage collected, the unfetched data is deleted too.

> reusing connection a connection with unfetched data can corrupt the following result data.

This is not currently a concern. The behavior of a query is not influenced by any previous query that was run on the same connection.

I think in practice you most likely don't need to worry about resetting connections. Though your use case might have some special properties of course that make this important.

**Comment by lucidfrontier45** [comment]
@danielmewes 

Thank you for advice.
I did not intend to reuse a cursor but a connection.

Just for clarity, in the following code

```python
conn = r.connect()

cur1 = r.table("test").run(conn)
next(cur1)
next(cur1)
next(cur1)

cur2 =  r.table("test").run(conn)
```

the `cur2` is not affected by `cur1` correct?

**Comment by danielmewes** [comment]
> I did not intend to reuse a cursor but a connection.

I understand. Sorry for causing confusion. The reason I talked about cursors was that cursors are the main thing that affect the state of a connection. A cursor is bound to a connection, and closing all cursors that use a given connection is pretty much equivalent to "resetting" the connection.

> the `cur2` is not affected by `cur1` correct?

That is correct. `cur2` will behave the same way regardless of `cur1`.

As long as `cur1` is still open and not closed or destructed, the server will keep a bit of bookkeeping information for `cur1` around. In an extreme case with a huge number of open cursors, this could lead to a significant increase in memory consumption on the server.
However as long as cursors are closed when they are no longer needed, I don't expect this to become an issue.

**Comment by lucidfrontier45** [comment]
OK, got it. Thank you.

**Comment by danielmewes** [comment]
I'm going to re-open this and put it in the "backlog" milestone. I think there might still be a use-case for resetting a connection (basically closing all cursors on it). I don't think it's a necessary feature for implementing a reliable connection pool, but I'd like to keep it on the list.
